Source/WebCore/animation/TimelineRange.cpp

```diff
@@ -43,11 +48,120 @@ bool SingleTimelineRange::isDefault(const CSSPrimitiveValue& value, Type type)
 }
 
+Length SingleTimelineRange::defaultValue(Type type)
+{
+    return type == Type::Start ? Length(0, LengthType::Percent) : Length(100, LengthType::Percent);
+}
+
 bool SingleTimelineRange::isOffsetValue(const CSSPrimitiveValue& value)
 {
     return value.isLength() || value.isPercentage() || value.isCalculatedPercentageWithLength();
 }
 
+TimelineRangeValue SingleTimelineRange::serialize() const
+{
+    return TimelineRangeOffset { CSSPrimitiveValue::create(valueID(name))->stringValue(), offset.isPercentOrCalculated() ? CSSNumericFactory::percent(offset.percent()) : CSSNumericFactory::px(offset.value()) };
+}
+
+static const std::optional<CSSToLengthConversionData> cssToLengthConversionData(RefPtr<Element> element)
+{
+    CheckedPtr elementRenderer = element->renderer();
+    if (!elementRenderer)
+        return std::nullopt;
+    CheckedPtr elementParentRenderer = elementRenderer->parent();
+    Ref document = element->document();
+    CheckedPtr documentElement = document->documentElement();
+    if (!documentElement)
+        return std::nullopt;
+
+    return CSSToLengthConversionData { elementRenderer->style(), documentElement->renderer() ? &documentElement->renderer()->style() : nullptr, elementParentRenderer ? &elementParentRenderer->style() : nullptr, document->renderView() };
+}
+
+Length SingleTimelineRange::lengthFoCSSValue(RefPtr<const CSSPrimitiveValue> value, RefPtr<Element> element)
+{
+    if (!value || value->isCalculated() || !element)
+        return { };
+    if (value->valueID() == CSSValueAuto)
+        return { };
+
+    auto conversionData = cssToLengthConversionData(element);
+    if (!conversionData) {
+        if (value->isPercentage())
+            return Length(value->resolveAsPercentageNoConversionDataRequired(), LengthType::Percent);
+        if (value->isPx())
+            return Length(value->resolveAsLengthNoConversionDataRequired(), LengthType::Fixed);
+        return { };
+    }
+
+    if (value->isLength())
+        return value->resolveAsLength<WebCore::Length>(*conversionData);
+
+    if (value->isPercentage())
+        return WebCore::Length(value->resolveAsPercentage(*conversionData), LengthType::Percent);
+
+    if (value->isCalculatedPercentageWithLength())
+        return WebCore::Length(value->cssCalcValue()->createCalculationValue(*conversionData, CSSCalcSymbolTable { }));
+
+    ASSERT_NOT_REACHED();
+    return { };
+}
+
+SingleTimelineRange SingleTimelineRange::range(const CSSValue& value, Type type, const Style::BuilderState* state, RefPtr<Element> element)
+{
+    if (RefPtr primitiveValue = dynamicDowncast<CSSPrimitiveValue>(value)) {
+        if (SingleTimelineRange::isOffsetValue(*primitiveValue)) {
+            // <length-percentage>
+            return { SingleTimelineRange::Name::Omitted, state ? Style::BuilderConverter::convertLength(*state, *primitiveValue) : lengthFoCSSValue(primitiveValue, element) };
```

Comment:
This is just stylistic, but I'd put the comment above the `if` so I don't need braces.

Reply:
Fixed.